DEVELOPING.rst

```diff
@@ -0,0 +1,238 @@
+system.
+
+The major components are:
+
+* :code:`renku.ui.cli`: The command line interface code using the ``click`` library.
+* :code:`renku.ui.rpc`: The JSON RPC server code used by the Renku platform.
+* :code:`renku.ui.api`: The user API allowing easy access to renku features
+  from users' Python scripts.
+* :code:`renku.command`: Wrappers following a Command Builder pattern (see
+  below) that is used to set up relevant context and post-execution code
+  for calls into ``renku.core``.
+* :code:`renku.infrastructure`: Contains implementations of interfaces for
+  interfacing with the outside world, such as the internal database and git.
+* :code:`renku.core`: The namespace for all business logic. Contains subfolders
+  for different topics, such as ``dataset``, ``workflow`` and so on.
+* :code:`renku.domain_model`: The domain models used within renku.
+* :code:`renku.data`: Contains non-code data files used by renku.
+
+Command Builder
+---------------
+
+Most of our business logic needs some context to be executed in, namely a Git
+repository/Renku project, the local database containing our metadata and other
+things such as requiring the repo to be clean, committing modified files after
+successful execution, checking that the project is up to date and so on.
+
+As this functionality is shared by many commands, we use a command builder to
+wrap business logic commands, allowing the context to be set up easily.
+
+The basic pattern is something like:
+
+.. code-block:: python
+
+   Command().command(_my_command).with_database(write=True).build().execute(arg1="a", arg2="")
+
+The above would wrap the function ``_my_command`` in a command builder, set up
+dependency injection so the metadata database is available in executed code
+using the ``with_database`` call and then finalize the command object using
+``build``. A finalized command cannot be further modified, all its settings are
+locked in after the call to ``build``.
+
+It can then be executed as many times as needed using its ``execute`` method,
+which takes care of setting up the relevant context (the database in this case)
+and then passed along its arguments to the wrapped ``_my_command``. Since we set
+``write=True``, this would also persist any changes made to metadata in the
+database after the command has finished.
+
+We usually don't call ``build`` in the ``renku.core`` namespace so callers of
+the command can still modify it with additional context if needed.
+
+
+Dependency Injection
+--------------------
+
+We use the inject_ library for dependency injection. DI is set up by the
+Command Builder for each command call.
+
+When an inner layer in the architecture needs to interact with objects from an
+outer layer, since dependencies aren't allowed to go outwards, there needs to
+be an interface in the inner layer that is the basis for injection and an
+implementation in the outer layer that gets injected.
+
+
+.. _inject: (https://pypi.org/project/Inject/)
```

Comment:
I've just noticed that this doesn't show up in GH or PyCharm. You need to remove parentheses for it to work correctly:
```
.. _inject: https://pypi.org/project/Inject/
```
It also happens in some other places and in `CONTRIBUTING.rst` as well.

Reply:
I'd copied that entry from CONTRIBUTING.rst and assumed it's fine :rofl: Went unnoticed for 2 years, good catch!